x-pack/plugins/security_solution/server/endpoint/services/actions/clients/endpoint/endpoint_actions_client.ts

```diff
@@ -0,0 +1,339 @@
+    }
+
+    return {
+      valid: validIds,
+      invalid: invalidIds,
+      allValid: invalidIds.length === 0,
+      hosts: foundEndpointHosts,
+    };
+  }
+
+  private async handleResponseAction<
+    TOptions extends ResponseActionsRequestBody = ResponseActionsRequestBody,
+    TResponse extends ActionDetails = ActionDetails,
+    TMethodOptions extends CommonResponseActionMethodOptions = CommonResponseActionMethodOptions
+  >(
+    command: ResponseActionsApiCommandNames,
+    actionReq: TOptions,
+    options?: TMethodOptions
+  ): Promise<TResponse> {
+    const agentIds = await this.checkAgentIds(actionReq.endpoint_ids);
+    const actionId = uuidv4();
+    const { hosts, ruleName, ruleId, error } = this.getMethodOptions<TMethodOptions>(options);
+    let actionError: string | undefined = error;
+
+    // Dispatch action to Endpoint using Fleet
+    if (!actionError) {
+      try {
+        await this.dispatchActionViaFleet({
+          actionId,
+          agents: agentIds.valid,
+          data: {
+            command,
+            comment: actionReq.comment,
+            parameters: actionReq.parameters as EndpointActionDataParameterTypes,
+          },
+        });
+      } catch (e) {
+        // If not in Automated mode, then just throw, else save the error and write
+        // it to the Endpoint Action request doc
+        if (!this.options.isAutomated) {
+          throw e;
+        }
+
+        actionError = e.message;
+      }
+    }
+
+    // Write action to endpoint index
+    await this.writeActionRequestToEndpointIndex({
+      ...actionReq,
+      error: actionError,
+      rule_id: ruleId,
+      rule_name: ruleName,
+      hosts,
+      actionId,
+      command,
+    });
+
+    // Update cases
+    await this.updateCases({
+      command,
+      actionId,
+      comment: actionReq.comment,
+      caseIds: actionReq.case_ids,
+      alertIds: actionReq.alert_ids,
+      hosts: actionReq.endpoint_ids.map((hostId) => {
+        return {
+          hostId,
+          hostname:
+            agentIds.hosts.find((host) => host.agent.id === hostId)?.host.hostname ??
+            hosts?.[hostId].name ??
+            '',
+        };
+      }),
+    });
+
+    return this.fetchActionDetails<TResponse>(actionId);
+  }
+
+  private async dispatchActionViaFleet({
+    actionId = uuidv4(),
+    agents,
+    data,
+  }: Pick<FleetActionRequest, 'agents'> & {
+    actionId?: string;
+    data: LogsEndpointAction['EndpointActions']['data'];
+  }): Promise<FleetActionRequest> {
+    const [messageSigningService, fleetActionsService] = await Promise.all([
+      this.options.endpointService.getMessageSigningService(),
+      this.options.endpointService.getFleetActionsClient(),
+    ]);
+    const fleetActionDoc: FleetActionRequest = {
+      '@timestamp': new Date().toISOString(),
+      input_type: 'endpoint',
+      type: 'INPUT_ACTION',
+      timeout: 300, // 5 minutes
```

Comment:
Maybe we should add a comment on the `timeout` property of `FleetActionRequest` that it is should accept values in seconds.

Reply:
Good point. I'll take a note on that and maybe add it along with a future PR. Don't want to trigger an approval from fleet just for a comment on a type